### DISCUSSION TITLE
How to set maximum memory > default 1GB for WASM?

### DISCUSSION THREAD

**Discussion by kjzl** [fzyzcjy/flutter_rust_bridge#1971] [Q&A] [answered]
Sadly I am exceeding the default memory limit of 1GB in my wasm build and naturally my first idea was to increase the limit...
I already tried...

setting in config.toml:
```
#[build]
rustflags = ["-C", "link-arg=--max-memory=4294967296"]
```

setting env var
`WASM_BINDGEN_THREADS_MAX_MEMORY="65536"`

directly modifying pkg/rust_lib_*.js's (maximum: 16384 -> 16384 * 4)
```
function __wbg_init_memory(imports, maybe_memory) {
    imports.wbg.memory = maybe_memory || new WebAssembly.Memory({initial:54,maximum:16384,shared:true});
}
```

At this point im desperate for anything to point me in the right direction, I can't seem to figure it out...

**Comment by fzyzcjy** [comment]
Hi, could you please show error logs about exceeding memory?

**Reply by kjzl** [reply]
I wonder why youre asking? I cant show right now, but does it matter? I'm processing large amounts of data in a yet not quite optimized fashion

**Reply by fzyzcjy** [reply]
Because I want to search that error message to see whether there is some quick suggestions. 

Btw it seems unrelated to flutter_rust_bridge core, so it would be great to also ask on related places as well.

**Reply by kjzl** [reply]
I understand that this probably does not have anything to do with frb, please bear with me for just a little bit
Maybe you can tell me why it is that when I run wasm-pack myself, without flutter_rust_bridge_codegen build-web, the memory limit will not be defined at all?
I did not really have tried anything with wasm before flutter_rust_bridge so Im not sure what else there is to it...

(The second line in the wasm file)
```
(module $rust_lib_frb_test.wasm
  (memory $wbg.memory (;0;) (export "memory") (import "wbg" "memory") 19 16384 shared)
  (func $wasm_bindgen::__wbindgen_object_clone_ref::h59261b61433d0e08 (;0;) (import "wbg" "__wbindgen_object_clone_ref") (param i32) (result i32))
```

will just be
```
(module $rust_lib_frb_test.wasm
  (func $wasm_bindgen::__wbindgen_object_clone_ref::h59261b61433d0e08 (;0;) (import "wbg" "__wbindgen_object_clone_ref") (param i32) (result i32))
```

**Reply by fzyzcjy** [reply]
> without flutter_rust_bridge_codegen build-web, the memory limit will not be defined at all?

Interesting! Indeed `build-web` does nothing but execute a few commands, i.e. a fairly simple wrapper script. So I guess you can have a look at those commands and see whether they define something special.

**Reply by kjzl** [reply, marked as answer]
Actually it seems I just fucked up by not exporting the env variables RUSTFLAGS and RUSTUP_TOOLCHAIN you define at https://github.com/fzyzcjy/flutter_rust_bridge/blob/193f5d0be3b4f16cab454d3e4e83c2bde9ed94ac/frb_dart/lib/src/cli/build_web/executor.dart#L162C6-L162C9

But in realizing you explicitly set the environment for wasm-pack I tested exporting RUSTFLAGS with the -C link-arg=--max-memory=4294967296 and then running wasm-pack, and yes, indeed this solved the problem. 
```
export RUSTUP_TOOLCHAIN="nightly"
export RUSTFLAGS="-C target-feature=+atomics,+bulk-memory,+mutable-globals -C link-arg=--max-memory=4294967296"
wasm-pack build -t no-modules -d /home/kjzl/Desktop/dev-projects/frb_test/web/pkg --no-typescript --out-name rust_lib_frb_test --dev rust -- -Z build-std=std,panic_abort
```

resulted in:
```
function __wbg_init_memory(imports, maybe_memory) {
    imports.wbg.memory = maybe_memory || new WebAssembly.Memory({initial:19,maximum:65536,shared:true});
}
```

**Reply by kjzl** [reply]
Maybe Its worth adding an option to provide own environment variables/extend the default ones in flutter_rust_bridge_codegen build-web or at this point add an option to define the memory limit?

Otherwise for anyone having the same issue, just use wasm-pack directly and set your own environment vars instead of using the frb build-web command

**Reply by fzyzcjy** [reply]
Good point! I guess the easiest way is to change things like

```dart
'RUSTUP_TOOLCHAIN': 'nightly',
```

into something like

```dart
'RUSTUP_TOOLCHAIN': 'nightly' + Platform.environment['RUSTUP_TOOLCHAIN'],
```

(maybe extract it to a separate function and properly handle null; the code above is just to express my thoughts)

Feel free to PR for this! Alternatively I will work on it in the next batch.
